// screener-india/lib/client/base.js
import axios from "axios";
import crypto from "crypto";
import { TtlCache } from "../utils/cache.js";
import { Throttle } from "../utils/throttle.js";
import { retryWithBackoff } from "../utils/retry.js";
const PARSER_VERSION = "1.0.0";
const DEFAULT_USER_AGENT = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";
export class BaseClient {
    http;
    cache;
    throttle;
    config;
    constructor(config = {}) {
        this.config = {
            baseUrl: config.baseUrl ?? "https://www.screener.in",
            cacheTtlMs: config.cacheTtlMs ?? 300_000,
            minIntervalMs: config.minIntervalMs ?? 200,
            maxRetries: config.maxRetries ?? 2,
            timeoutMs: config.timeoutMs ?? 20_000,
            proxyUrl: config.proxyUrl ?? "",
            cookies: config.cookies ?? "",
        };
        const defaultHeaders = {
            "User-Agent": DEFAULT_USER_AGENT,
            Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9",
        };
        if (this.config.cookies) {
            defaultHeaders["Cookie"] = this.config.cookies;
        }
        this.http = axios.create({
            timeout: this.config.timeoutMs,
            headers: defaultHeaders,
        });
        this.cache = new TtlCache(this.config.cacheTtlMs);
        this.throttle = new Throttle(this.config.minIntervalMs);
    }
    cacheKey(url) {
        return crypto.createHash("sha256").update(url).digest("hex");
    }
    meta(url) {
        return {
            sourceUrl: url,
            fetchedAt: new Date().toISOString(),
            parserVersion: PARSER_VERSION,
        };
    }
    async fetchHtml(url, scope = "default") {
        const key = this.cacheKey(url);
        const cached = this.cache.get(key);
        if (cached)
            return cached;
        await this.throttle.wait(scope);
        const html = await retryWithBackoff(async () => {
            const res = await this.http.get(url, {
                responseType: "text",
                maxRedirects: 5,
            });
            // Detect login-wall: screener.in redirects protected pages to /register/ or /login/
            const finalUrl = res.request?.res?.responseUrl ?? "";
            if (finalUrl.includes("/register") || finalUrl.includes("/login")) {
                throw new Error(`LOGIN_REQUIRED: ${url} requires a screener.in account`);
            }
            return res.data;
        }, this.config.maxRetries);
        this.cache.set(key, html);
        return html;
    }
    companyUrl(symbol, mode) {
        const base = `${this.config.baseUrl}/company/${symbol.toUpperCase()}`;
        return mode === "consolidated" ? `${base}/consolidated/` : `${base}/`;
    }
}
